### DISCUSSION TITLE
How to save and restore the folded-unfolded state of the map

### DISCUSSION THREAD

**Discussion by Memmaleph** [freeplane/freeplane#1606] [Help]
Hi, 

I have a map with many different notes on various subjects and I'd like to be able to easily restore the state of the folded and unfolded nodes. Is there a simple way to do so?

For example, after searching or filtering, the (un)folded state of the nodes might change and I'm not always sure of what was the exact structure before. 

I'm not sure, but I think that before, I could just close FreePlane without saving and reopening it. But I feel that this behavior has changed. Did I missed something?

**Comment by EdoFro** [comment]
> I'm not sure, but I think that before, I could just close FreePlane without saving and reopening it. But I feel that this behavior has changed. Did I missed something?

look at this preference's parameter: `Preferences…->Environment->Save->Save folding`

Also if you use multiple views for the same map, each one of them can show different folding states for the nodes. The saved folding state for a node is the last folding/unfolding done to it. It doesn't matter in which view it was. 
In other words. Freeplane doesn't save the folding state of one particular view (the active one, or the last closed), it save for each node the last folding change done by the user.
I find this can be tricky.

**Comment by EdoFro** [comment]
Here is also a little script that stores the folding state information in an attribute in the root node of the map:

```groovy
def foldingAttributeName = 'foldedIDs'

def mapa = node.mindMap
def rootNode = mapa.root

def nowFoldedNodes  = [] + rootNode.find{it.folded}
println 'nowFoldedNodes*.id:\n\t' + nowFoldedNodes*.id

def foldedIDs

if(!rootNode[foldingAttributeName]){
    println 'storing folding states'
    foldedIDs = nowFoldedNodes.collect{it.id}
    rootNode[foldingAttributeName] = foldedIDs
} else {
    println 'applying stored folding states'
    foldedIDs = rootNode[foldingAttributeName].drop(1).dropRight(1).split(', ').findAll()
    
    def oldFoldedNodes = foldedIDs.collect{mapa.node(it)}
    println oldFoldedNodes
    println 'oldFoldedNodes*.id:\n\t' + oldFoldedNodes*.id
    
    //actions:
    //fold unfolded
    println 'currently unfolded that will be folded (old - now):\n\t' + (oldFoldedNodes - nowFoldedNodes)*.id
    (oldFoldedNodes - nowFoldedNodes).each{it.folded = true}
    
    //unfold folded
    println 'currently folded that will be unfolded (now - old):\n\t' + (nowFoldedNodes - oldFoldedNodes)*.id
    (nowFoldedNodes - oldFoldedNodes).each{it.folded = false}

    rootNode[foldingAttributeName] = null
}
```

hope it helps,

edo
